benchexec/tools/coccinelle.py

```diff
@@ -0,0 +1,64 @@
+# This file is part of BenchExec, a framework for reliable benchmarking:
+# https://github.com/sosy-lab/benchexec
+#
+# SPDX-FileCopyrightText: 2007-2020 Dirk Beyer <https://www.sosy-lab.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import os
+import subprocess
+
+import benchexec.result as result
+from benchexec.tools.template import ToolNotFoundException, BaseTool2
+
+
+class Tool(BaseTool2):
+    """
+    Tool info for Coccinelle
+    (https://github.com/coccinelle/coccinelle).
+    """
+
+    REQUIRED_PATHS = ["standard.h" "ocaml" "standard.iso" "ocaml"]
+
+    def name(self):
+        return "Coccinelle"
+
+    def executable(self, tool_locator):
+        exe = tool_locator.find_executable("spatch")
+        dir_name = os.path.dirname(exe)
+        logging.debug("Looking in %s for spatch", dir_name)
+        for _, dir_names, file_names in os.walk(dir_name):
+            if "spatch" in file_names:
+                return exe
+            break
+        msg = (
+            f"ERROR: Did find a spatch in {os.path.dirname(exe)} "
+            f"but no 'spatch' directory besides it"
+        )
+        raise ToolNotFoundException(msg)
+
+    def version(self, executable):
+        return self._version_from_tool(executable)
+
+    def cmdline(self, executable, options, task, resource_limits):
+        return [executable] + options + list(task.input_files) + [task.property_file]
+
+    def program_files(self, executable):
+        paths = self.REQUIRED_PATHS
+        return [executable] + self._program_files_from_executable(executable, paths)
```

Comment:
There seems to be no reason to have this method, the default implementation does the same.